File: java-shim/src/main/java/io/skillgate/SkillGateClient.java
package io.skillgate;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.Instant;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Thread-safe HTTP client for the SkillGate runtime sidecar.
 *
 * <pre>{@code
 * SkillGateClient client = new SkillGateClient(SkillGateClient.Config.fromEnv());
 * DecisionRecord decision = client.decide(invocation);
 * if (!"ALLOW".equals(decision.decision())) {
 *     throw new RuntimeException("Denied: " + decision.decisionCode());
 * }
 * }</pre>
 */
public final class SkillGateClient implements AutoCloseable {

    // ---- Models -------------------------------------------------------------

    /** Actor invoking the tool. */
    public record Actor(
        @JsonProperty("type") String type,
        @JsonProperty("id") String id,
        @JsonProperty("workspace_id") String workspaceId,
        @JsonProperty("session_id") String sessionId
    ) {}

    /** Agent metadata. */
    public record Agent(
        @JsonProperty("name") String name,
        @JsonProperty("version") String version,
        @JsonProperty("framework") String framework,
        @JsonProperty("trust_tier") String trustTier
    ) {}

    /** Tool metadata. */
    public record Tool(
        @JsonProperty("name") String name,
        @JsonProperty("provider") String provider,
        @JsonProperty("capabilities") List<String> capabilities,
        @JsonProperty("risk_class") String riskClass
    ) {}

    /** Tool call parameters and referenced resources. */
    public record ToolRequest(
        @JsonProperty("params") Map<String, Object> params,
        @JsonProperty("resource_refs") List<String> resourceRefs
    ) {
        public ToolRequest() { this(Map.of(), List.of()); }
    }

    /** Execution environment metadata. */
    public record ExecutionContext(
        @JsonProperty("repo") String repo,
        @JsonProperty("environment") String environment,
        @JsonProperty("data_classification") String dataClassification,
        @JsonProperty("network_zone") String networkZone
    ) {}

    /** Canonical enforcement request payload. */
    public record ToolInvocation(
        @JsonProperty("invocation_id") String invocationId,
        @JsonProperty("timestamp") Instant timestamp,
        @JsonProperty("actor") Actor actor,
        @JsonProperty("agent") Agent agent,
        @JsonProperty("tool") Tool tool,
        @JsonProperty("request") ToolRequest request,
        @JsonProperty("context") ExecutionContext context
    ) {}

    /** Budget snapshot for a single capability. */
    @JsonIgnoreProperties(ignoreUnknown = true)
    public record BudgetStatus(
        @JsonProperty("remaining") long remaining,
        @JsonProperty("limit") long limit
    ) {}

    /** Signed attestation evidence. */
    @JsonIgnoreProperties(ignoreUnknown = true)
    public record DecisionEvidence(
        @JsonProperty("hash") String hash,
        @JsonProperty("signature") String signature,
        @JsonProperty("key_id") String keyId
    ) {}

    /** Enforcement decision returned by the sidecar. */
    @JsonIgnoreProperties(ignoreUnknown = true)
    public record DecisionRecord(
        @JsonProperty("invocation_id") String invocationId,
        /** "ALLOW" | "DENY" | "FAIL" | "REQUIRE_APPROVAL" */
        @JsonProperty("decision") String decision,
        @JsonProperty("decision_code") String decisionCode,
        @JsonProperty("reason_codes") List<String> reasonCodes,
        @JsonProperty("policy_version") String policyVersion,
        @JsonProperty("budgets") Map<String, BudgetStatus> budgets,
        @JsonProperty("evidence") DecisionEvidence evidence,
        @JsonProperty("degraded") boolean degraded,
        @JsonProperty("entitlement_version") String entitlementVersion,
        @JsonProperty("license_mode") String licenseMode
    ) {}

    // ---- Exception ----------------------------------------------------------

    /** Thrown when the sidecar is unreachable and fail-open is disabled. */
    public static final class EnforcerUnavailableException extends IOException {
        public EnforcerUnavailableException(String message) { super(message); }
        public EnforcerUnavailableException(String message, Throwable cause) { super(message, cause); }
    }

    // ---- Config -------------------------------------------------------------

    /**
     * Client configuration.
     *
     * @param sidecarUrl  Sidecar base URL. Default: {@code http://localhost:8910}.
     * @param timeoutMs   Per-request timeout in milliseconds. Default: 50.
     * @param failOpen    When true, return a degraded ALLOW on sidecar failure.
     * @param slt         Session License Token for the Authorization header.
     */
    public record Config(String sidecarUrl, int timeoutMs, boolean failOpen, String slt) {

        /** Build configuration from environment variables with production-safe defaults. */
        public static Config fromEnv() {
            String url = System.getenv("SKILLGATE_SIDECAR_URL");
            if (url == null || url.isBlank()) url = "http://localhost:8910";
            String slt = System.getenv("SKILLGATE_SLT");
            return new Config(url, 50, false, slt);
        }
    }

    // ---- Client impl --------------------------------------------------------

    private final Config cfg;
    private final HttpClient http;
    private final ObjectMapper mapper;

    public SkillGateClient(Config cfg) {
        this.cfg = cfg;
        this.http = HttpClient.newBuilder()
            .connectTimeout(Duration.ofMillis(cfg.timeoutMs()))
            .build();
        this.mapper = new ObjectMapper()
            .registerModule(new JavaTimeModule())
            .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE);
    }

    private Map<String, String> authHeaders() {
        if (cfg.slt() != null && !cfg.slt().isBlank()) {
            return Map.of("Authorization", "Bearer " + cfg.slt());
        }
        return Map.of();
    }

    private DecisionRecord degradedAllow(String invocationId) {
        return new DecisionRecord(
            invocationId, "ALLOW", "SG_ALLOW_DEGRADED_AUDIT_ASYNC",
            List.of("enforcer_unavailable_fail_open"), "unknown",
            Map.of(), new DecisionEvidence("", "", ""),
            true, "unknown", "offline"
        );
    }

    /**
     * Send a {@link ToolInvocation} to the sidecar for an enforcement decision.
     *
     * @throws EnforcerUnavailableException if sidecar is unreachable and fail-open is false.
     */
    public DecisionRecord decide(ToolInvocation invocation) throws IOException, InterruptedException {
        Map<String, Object> body = new HashMap<>();
        body.put("invocation_id", invocation.invocationId());
        body.put("tool_invocation", invocation);

        String json = mapper.writeValueAsString(body);

        HttpRequest.Builder reqBuilder = HttpRequest.newBuilder()
            .uri(URI.create(cfg.sidecarUrl() + "/v1/decide"))
            .header("Content-Type", "application/json")
            .timeout(Duration.ofMillis(cfg.timeoutMs()))
            .POST(HttpRequest.BodyPublishers.ofString(json));

        authHeaders().forEach(reqBuilder::header);

        try {
            HttpResponse<String> response = http.send(reqBuilder.build(), HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() >= 400) {
                throw new IOException("Sidecar returned " + response.statusCode() + ": " + response.body());
            }
            return mapper.readValue(response.body(), DecisionRecord.class);
        } catch (IOException e) {
            if (cfg.failOpen()) return degradedAllow(invocation.invocationId());
            throw new EnforcerUnavailableException("SkillGate sidecar unreachable (fail-closed)", e);
        }
    }

    /**
     * Register or update a tool AI-BOM in the sidecar registry.
     * Best-effort — returns {@code false} on any connectivity failure.
     */
    public boolean registerTool(String toolName, Map<String, Object> metadata) {
        try {
            String json = mapper.writeValueAsString(metadata);
            HttpRequest.Builder reqBuilder = HttpRequest.newBuilder()
                .uri(URI.create(cfg.sidecarUrl() + "/v1/registry/" + toolName))
                .header("Content-Type", "application/json")
                .timeout(Duration.ofMillis(cfg.timeoutMs()))
                .PUT(HttpRequest.BodyPublishers.ofString(json));
            authHeaders().forEach(reqBuilder::header);
            HttpResponse<Void> resp = http.send(reqBuilder.build(), HttpResponse.BodyHandlers.discarding());
            return resp.statusCode() >= 200 && resp.statusCode() < 300;
        } catch (IOException | InterruptedException e) {
            Thread.currentThread().interrupt();
            return false;
        }
    }

    /**
     * Returns {@code true} if the sidecar is reachable and healthy.
     */
    public boolean isHealthy() {
        try {
            HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(cfg.sidecarUrl() + "/v1/health"))
                .timeout(Duration.ofMillis(cfg.timeoutMs()))
                .GET().build();
            HttpResponse<Void> resp = http.send(req, HttpResponse.BodyHandlers.discarding());
            return resp.statusCode() == 200;
        } catch (IOException | InterruptedException e) {
            Thread.currentThread().interrupt();
            return false;
        }
    }

    @Override
    public void close() {
        // HttpClient is closeable in Java 21+; nothing to do for 17
    }
}

File: dotnet-shim/SkillGate.Client/SkillGateClient.cs
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace SkillGate;

// ---- Models -----------------------------------------------------------------

/// <summary>Actor invoking the tool.</summary>
public sealed record Actor(
    [property: JsonPropertyName("type")] string Type,
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("workspace_id")] string WorkspaceId,
    [property: JsonPropertyName("session_id")] string SessionId
);

/// <summary>Agent metadata.</summary>
public sealed record Agent(
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("version")] string Version,
    [property: JsonPropertyName("framework")] string Framework,
    [property: JsonPropertyName("trust_tier")] string TrustTier
);

/// <summary>Tool metadata.</summary>
public sealed record Tool(
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("provider")] string Provider,
    [property: JsonPropertyName("capabilities")] IReadOnlyList<string> Capabilities,
    [property: JsonPropertyName("risk_class")] string RiskClass
);

/// <summary>Tool call parameters and referenced resources.</summary>
public sealed record ToolRequest(
    [property: JsonPropertyName("params")] IReadOnlyDictionary<string, object?> Params,
    [property: JsonPropertyName("resource_refs")] IReadOnlyList<string> ResourceRefs
)
{
    public ToolRequest() : this(new Dictionary<string, object?>(), Array.Empty<string>()) { }
}

/// <summary>Execution environment metadata.</summary>
public sealed record ExecutionContext(
    [property: JsonPropertyName("repo")] string Repo,
    [property: JsonPropertyName("environment")] string Environment,
    [property: JsonPropertyName("data_classification")] string DataClassification,
    [property: JsonPropertyName("network_zone")] string NetworkZone
);

/// <summary>Canonical enforcement request payload.</summary>
public sealed record ToolInvocation(
    [property: JsonPropertyName("invocation_id")] string InvocationId,
    [property: JsonPropertyName("timestamp")] DateTimeOffset Timestamp,
    [property: JsonPropertyName("actor")] Actor Actor,
    [property: JsonPropertyName("agent")] Agent Agent,
    [property: JsonPropertyName("tool")] Tool Tool,
    [property: JsonPropertyName("request")] ToolRequest Request,
    [property: JsonPropertyName("context")] ExecutionContext Context
);

/// <summary>Budget snapshot for a capability.</summary>
public sealed record BudgetStatus(
    [property: JsonPropertyName("remaining")] long Remaining,
    [property: JsonPropertyName("limit")] long Limit
);

/// <summary>Signed attestation evidence.</summary>
public sealed record DecisionEvidence(
    [property: JsonPropertyName("hash")] string Hash,
    [property: JsonPropertyName("signature")] string Signature,
    [property: JsonPropertyName("key_id")] string KeyId
);

/// <summary>Enforcement decision returned by the sidecar.</summary>
public sealed record DecisionRecord(
    [property: JsonPropertyName("invocation_id")] string InvocationId,
    /// <summary>"ALLOW" | "DENY" | "FAIL" | "REQUIRE_APPROVAL"</summary>
    [property: JsonPropertyName("decision")] string Decision,
    [property: JsonPropertyName("decision_code")] string DecisionCode,
    [property: JsonPropertyName("reason_codes")] IReadOnlyList<string> ReasonCodes,
    [property: JsonPropertyName("policy_version")] string PolicyVersion,
    [property: JsonPropertyName("budgets")] IReadOnlyDictionary<string, BudgetStatus> Budgets,
    [property: JsonPropertyName("evidence")] DecisionEvidence Evidence,
    [property: JsonPropertyName("degraded")] bool Degraded,
    [property: JsonPropertyName("entitlement_version")] string EntitlementVersion,
    [property: JsonPropertyName("license_mode")] string LicenseMode
);

// ---- Exceptions -------------------------------------------------------------

/// <summary>Thrown when the sidecar is unreachable and FailOpen is false.</summary>
public sealed class EnforcerUnavailableException(string message, Exception? inner = null)
    : Exception(message, inner);

// ---- Config -----------------------------------------------------------------

/// <summary>Client configuration.</summary>
public sealed record SkillGateConfig(
    /// <summary>Sidecar base URL. Default: http://localhost:8910.</summary>
    string SidecarUrl,
    /// <summary>Per-request timeout. Default: 50 ms.</summary>
    TimeSpan Timeout,
    /// <summary>When true, return a degraded ALLOW on sidecar failure.</summary>
    bool FailOpen,
    /// <summary>Session License Token for the Authorization header.</summary>
    string? Slt
)
{
    /// <summary>Build config from environment variables with production-safe defaults.</summary>
    public static SkillGateConfig FromEnv()
    {
        var url = Environment.GetEnvironmentVariable("SKILLGATE_SIDECAR_URL");
        if (string.IsNullOrWhiteSpace(url)) url = "http://localhost:8910";
        var slt = Environment.GetEnvironmentVariable("SKILLGATE_SLT");
        return new SkillGateConfig(url, TimeSpan.FromMilliseconds(50), false, slt);
    }
}

// ---- Client -----------------------------------------------------------------

/// <summary>
/// Thread-safe HTTP client for the SkillGate runtime sidecar.
/// </summary>
public sealed class SkillGateClient : IDisposable
{
    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    private readonly SkillGateConfig _cfg;
    private readonly HttpClient _http;

    public SkillGateClient(SkillGateConfig cfg)
    {
        _cfg = cfg;
        _http = new HttpClient { Timeout = cfg.Timeout };
        if (!string.IsNullOrWhiteSpace(cfg.Slt))
            _http.DefaultRequestHeaders.Authorization =
                new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", cfg.Slt);
    }

    private static DecisionRecord DegradedAllow(string invocationId) => new(
        invocationId, "ALLOW", "SG_ALLOW_DEGRADED_AUDIT_ASYNC",
        ["enforcer_unavailable_fail_open"], "unknown",
        new Dictionary<string, BudgetStatus>(),
        new DecisionEvidence("", "", ""),
        true, "unknown", "offline"
    );

    /// <summary>
    /// Send a <see cref="ToolInvocation"/> to the sidecar for an enforcement decision.
    /// </summary>
    /// <exception cref="EnforcerUnavailableException">
    /// Sidecar unreachable and <see cref="SkillGateConfig.FailOpen"/> is false.
    /// </exception>
    public async Task<DecisionRecord> DecideAsync(
        ToolInvocation invocation,
        CancellationToken cancellationToken = default)
    {
        var body = new { invocation_id = invocation.InvocationId, tool_invocation = invocation };

        try
        {
            var response = await _http.PostAsJsonAsync(
                $"{_cfg.SidecarUrl}/v1/decide", body, JsonOpts, cancellationToken);
            response.EnsureSuccessStatusCode();
            var record = await response.Content.ReadFromJsonAsync<DecisionRecord>(JsonOpts, cancellationToken);
            return record!;
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException or TimeoutException)
        {
            if (_cfg.FailOpen) return DegradedAllow(invocation.InvocationId);
            throw new EnforcerUnavailableException("SkillGate sidecar unreachable (fail-closed)", ex);
        }
    }

    /// <summary>
    /// Register or update a tool AI-BOM in the sidecar registry.
    /// Best-effort — returns false on any connectivity failure.
    /// </summary>
    public async Task<bool> RegisterToolAsync(
        string toolName,
        Dictionary<string, object?> metadata,
        CancellationToken cancellationToken = default)
    {
        try
        {
            var response = await _http.PutAsJsonAsync(
                $"{_cfg.SidecarUrl}/v1/registry/{toolName}", metadata, JsonOpts, cancellationToken);
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }

    /// <summary>Returns true if the sidecar is reachable and healthy.</summary>
    public async Task<bool> IsHealthyAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            var response = await _http.GetAsync($"{_cfg.SidecarUrl}/v1/health", cancellationToken);
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }

    public void Dispose() => _http.Dispose();
}
